plugins/rts54hub/fu-rts54hub-rtd21xx-foreground.c

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (C) 2021 Realtek Corporation
+ * Copyright (C) 2021 Ricky Wu <ricky_wu@realtek.com> <spring1527@gmail.com>
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-chunk.h"
+
+#include "fu-rts54hub-rtd21xx-foreground.h"
+#include "fu-rts54hub-device.h"
+
+struct _FuRts54hubRtd21xxForeground
+{
+	FuRts54hubRtd21xxDevice	 parent_instance;
+};
+
+G_DEFINE_TYPE (FuRts54hubRtd21xxForeground, fu_rts54hub_rtd21xx_foreground, FU_TYPE_RTS54HUB_RTD21XX_DEVICE)
+
+#define ISP_DATA_BLOCKSIZE		256
+#define ISP_PACKET_SIZE			257
+
+typedef enum {
+	ISP_CMD_ENTER_FW_UPDATE		= 0x01,
+	ISP_CMD_GET_PROJECT_ID_ADDR	= 0x02,
+	ISP_CMD_SYNC_IDENTIFY_CODE	= 0x03,
+	ISP_CMD_GET_FW_INFO		= 0x04,
+	ISP_CMD_FW_UPDATE_START		= 0x05,
+	ISP_CMD_FW_UPDATE_ISP_DONE	= 0x06,
+	ISP_CMD_FW_UPDATE_EXIT		= 0x07,
+	ISP_CMD_FW_UPDATE_RESET		= 0x08,
+} IspCmd;
+
+static gboolean
+fu_rts54hub_rtd21xx_ensure_version_unlocked (FuRts54hubRtd21xxForeground *self,
+					     GError **error)
+{
+	guint8 buf_rep[7] = { 0x00 };
+	guint8 buf_req[] = { ISP_CMD_GET_FW_INFO };
+	guint8 buf[] = { ISP_CMD_FW_UPDATE_RESET };
+	g_autofree gchar *version = NULL;
+
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (FU_RTS54HUB_RTD21XX_DEVICE (self),
+						   UC_ISP_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   buf_req, sizeof(buf_req),
+						   error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+
+	/* wait for device ready */
+	g_usleep(300000);
+	if (!fu_rts54hub_rtd21xx_device_i2c_read (FU_RTS54HUB_RTD21XX_DEVICE (self),
+						  UC_ISP_SLAVE_ADDR,
+						  0x00,
+						  buf_rep, sizeof(buf_rep),
+						  error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+	/* set version */
+	version = g_strdup_printf ("%u.%u", buf_rep[1], buf_rep[2]);
+	fu_device_set_version (FU_DEVICE (self), version);
+
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (FU_RTS54HUB_RTD21XX_DEVICE (self),
+						   UC_ISP_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   buf, sizeof(buf),
+						   error)) {
+		g_prefix_error (error, "failed to ISP_CMD_FW_UPDATE_RESET: ");
+		return FALSE;
+	}
+
```

Comment:
RESET Command Should not send here.

Reply:
Command  Define is wrong. ISP_CMD_FW_UPDATE_EXIT Value shoud be  0x08, ISP_CMD_FW_UPDATE_RESET value	shoud be 0x07.  Read version do't need reset.,use exit.  Foreground FW Upadte user reset. 
If the exit  command send  here, it will encountered an error in attach function. Because the device slave address is change back to 0x6A.

Reply:
I'm not sure I completely follow. Could you do a pull request with the target to be wip/hughsie/rts54hub and I'll merge that. Thanks.

Reply:
I've merged the fixes; can you test this branch again and if it works for you then I think it's good to merge.